---
title: "scRNA-seq template"
author: "Stephanie Strohbuecker"
subtitle: Prepare the directory structure for the single cell analysis
output:
  html_document:
    df_print: paged
---

```{r global_options, include=FALSE}
knitr::opts_chunk$set(echo = FALSE,
                      eval.after = "fig.cap",
                      fig.width = 12)
```

# onDemand RStudio:
When using onDemand RStudio, the proxy settings need to be also set in the `.Renviron`
```{bash, eval=FALSE}
CURL_CA_BUNDLE=/etc/ssl/certs/ca-bundle.crt
http_proxy=http://10.28.7.200:3128
https_proxy=http://10.28.7.200:3128
ftp_proxy=http://10.28.7.200:3128
```

Also when requesting an R session via onDemand please as for this additional module
```{bash, eval=FALSE}
CAMP_proxy/0.1
```

# Git and GitHub

Note on setting up a github repository for this analysis: We followed the tutorial [here](https://happygitwithr.com/rstudio-git-github.html)

```{r}
RMD_file_name <- "0-1.set_up_analysis_environment"
```


# Using `renv`

We are using `renv` with this project to capture the packages used. For more information on `renv` please refer to this [vignette](https://rstudio.github.io/renv/articles/renv.html).

For setting up `renv` on CAMP using a shared cache for `renv`, the following lines need to be added to the `.Renviron` file in the home directory:
```{bash, eval=FALSE}
RENV_PATHS_ROOT=/camp/stp/babs/working/software/renv
RENV_PATHS_LOCAL=/camp/stp/babs/working/software/renv/babs_packages
```


```{r, eval = FALSE}
# Only required if the project was initialized without renv
renv::init()
```


```{r}
# update packages to the most recent version before starting the project
# update, without prompts for permission/clarification
update.packages(ask = FALSE)
```

```{r}
# update the renv.lockfile to reflect the latest status of the libraries available for this analysis
renv::snapshot()
```


Please note that package installation should be done wherever possible via `renv::install()`
```{r, eval=FALSE}
# for CRAN packages:
renv::install("package_name")
# for Bioconductor packages
renv::install("bioc::package_name")
# for packages from GitHub
renv::install("path/to/github/repro")
# for packages on the local path
renv::install("~/path/to/package")
```


# Libraries

```{r}
library(tidyverse)
library(here)
```

# Setting up the directory structure

According to a [blog post from Jenny Bryant](https://www.tidyverse.org/blog/2017/12/workflow-vs-script/) the aim for this project directory is that it is self-contained, i.e. the project folder can be moved and everything will still work.
We use the `here()` function to build paths when reading or writing files. The `here package` uses the `.Rproj` file to identify the top-level directory from which relative paths are specified. The `.Rproj` file is left by RStudio.
(Note when working without RStudio launch the R process from the project's toplevel directory. There are other ways to mark the toplevel directory for here when working outside RStudio, e.g. adding an empty file called `.here` or a `DESCRIPTION` file to the toplevel directory.)

The directory structure is modeled on an R package used as a "Research compendium" as described by [Marwick et al. 2018](https://peerj.com/preprints/3192v2/).

The basic project structure contains the following directories and files

* R.proj file
* DESCRIPTION file
* README.md file
* LICENCE file (optional, only necessary for publication)
* NAMESPACE (autogenerated file that exports R functions for repeated use)
* R/
    * directory containing R functions (generate them via devtools::ure_r(), devtools::load_all(), and devtools::document() (see [R packages](https://r-pkgs.org/index.html) for more details)
* man/
    * directory containg documentation for the functions in the R/ directory
* templates/
    * directory containing templates, e.g. `.csl` files to set citation and reference list style, etc.
* data/
    * with subdirectories containing 
        * raw data files
        * derived data files
        * external data files, e.g. further information provided by the scientist
* figures/
    * directory containing figures generated during the analysis
* vignettes/
    * directory containing final reports (based on the analysis files)
* analysis/
    * individual `.Rmd` files with the individual analysis steps carried out

Note that not all the folders need to be generated for all projects. The following will generate the minimal directory structure

## Timekeeping file

Before proceeding we need to ensure that the expected `.babs` timekeepiing file is available in the toplevel directory (the one marked by the `.Rproj` file).
For more information on how to claim/generate the appropriate file please refer [Gavin Kelly's timekeeping wiki](https://github.com/crickbabs/BABS-Docs/wiki/Time).

We first need to write the DESCRIPTION file. We base the DESCRIPTION file on the information available in the `.babs` file:
```{r}
proj_info <- read_tsv("./.babs", col_names = FALSE) %>% 
  separate(col = X1, into = c("descriptor", "specification"), sep =":")

min_R_version <- str_c("R", "(>= ", R.version$major, ".", R.version$minor, ")", sep = "")

description <-
  str_c(
    str_to_title(proj_info$specification[proj_info$descriptor == "Type"]),
    " analysis for",
    proj_info$specification[proj_info$descriptor == "Scientist"],
    " with the project title" ,
    proj_info$specification[proj_info$descriptor == "Scititle"],
    ". For more information please refer to the project proposal .docx in ./data/raw_data/project_docs.",
    sep = ""
  )
```


```{r, eval=FALSE}
usethis::use_description(
  fields = list(
    `Authors@R` = c(
      utils::person(
        "Stephanie",
        "Strohbuecker",
        email = "stephanie.strohbuecker@crick.ac.uk",
        role = c("aut", "cre"),
        comment = c(ORCID = "0000-0002-9781-6879")
      ),
      utils::person("BABS",
                    email = "bioinformatics@crick.ac.uk",
                    role = c("aut", "cre"))
    ),
    Package = proj_info %>%
      filter(descriptor == "Lims") %>%
      pull(specification),
    Title = proj_info %>%
      filter(descriptor == "Scititle") %>%
      pull(specification),
    Depends = min_R_version,
    Description = description,
    License = "GLP-3",
    LazyData = "TRUE"
  ),
  check_name = FALSE
)
```

After having generated the DESCRIPTION file, we can use the functions that are part of the `devtools` universe (i.e. `usethis`, `pkgload`).

With the following code chunks we can generate functions with appropriate documentation for usage within this project.
```{r, eval=FALSE}
# generate a file for a function to be used project-wide
# The code for the function needs to be copied into the newly generated file from the command below
usethis::use_r("generate_directories") 
```

```{r, eval=FALSE}
# for documenting the function insert cursor in the function definition in the appropriate .R file, go to Code> Insert roxygen skeleton
# fill out the skeleton with the appropriate information
# finish manual creation for that function with
devtools::document()
```


```{r}
# load_all() makes the functions in the R/ directory available for usage
pkgload::load_all()
```


## Generate the basic directory structure for this project

```{r}
map(c("data/raw_data/project_docs", "data/derived_data", "figures", "vignettes", "analysis/scripts", "analysis/logs"), ~{
  generate_directories(.x)
})
```

Move files provided by the scientist in the appropriate directory
```{r}
filesstrings::move_files(list.files(".", pattern = ".docx$"), destinations = "./data/raw_data/project_docs/")
filesstrings::move_files(list.files(".", pattern = ".xlsx$"), destinations = "./data/raw_data/project_docs/")
```

```{r}
filesstrings::move_files(list.files(".", pattern = ".csv$"), destinations = "./data/raw_data/project_docs/")
```

```{r}
proj_info <- proj_info %>% 
  mutate(specification = trimws(specification))
raw_data_base <- str_c(
  "/camp/stp/babs/inputs/sequencing/data",
  proj_info$specification[proj_info$descriptor == "Lab"],
  proj_info$specification[proj_info$descriptor == "Scientist"],
  proj_info$specification[proj_info$descriptor == "Lims"],
  sep = "/"
)
```


```{r}
file.copy(list.files(raw_data_base, pattern = ".csv$", full.names = TRUE), "./data/raw_data/project_docs/")
```

